imagefun/imagefun.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"image"
+	"image/color"
+	"image/png"
+	"log"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+)
+
+// The agenda:
+
+// struct
+// errors
+// panics (Aaaaaaaaa!)
+// arrays & slices
+// maps
+// pointers
+
+const (
+	widthDefault    = 200
+	heightDefault   = 200
+	fileNameDefault = "image.png"
+)
+
+var (
+	colors = map[int]*color.RGBA{
+		// RGBa: RED, GREEN, BLUE, ALPHA ch. (transparency)
+		0:  {R: 100, G: 200, B: 200, A: 0xff},
+		1:  {R: 70, G: 70, B: 21, A: 0xff},
+		2:  {R: 207, G: 70, B: 110, A: 0xff},
+		3:  {R: 78, G: 70, B: 207, A: 0xff},
+		4:  {R: 207, G: 205, B: 70, A: 0xff},
+		5:  {R: 177, G: 37, B: 180, A: 0xff},
+		6:  {R: 225, G: 39, B: 232, A: 0xff},
+		7:  {R: 255, G: 0, B: 0, A: 0xff},
+		8:  {R: 0, G: 255, B: 0, A: 0xff},
+		9:  {R: 0, G: 0, B: 255, A: 0xff},
+		10: {R: 255, G: 128, B: 0, A: 0xff},
+		11: {R: 255, G: 0, B: 127, A: 0xff},
+	}
+	err   error
+	lines []string
+)
+
+type imgOptions struct {
+	width    int
+	height   int
+	colors   map[int]*color.RGBA
+	fileName string
+	content  [][]*color.RGBA
+}
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+func main() {
+
+	var w *int = flag.Int("w", widthDefault, "image width in pixels")
+	var h *int = flag.Int("h", heightDefault, "image height in pixels")
+	var fn *string = flag.String("f", fileNameDefault, "file name")
+	var isFlag *bool = flag.Bool("flag", false, "draw flag image based on user input")
+	flag.Parse()
+
+	var options imgOptions = imgOptions{
+		width:    *w,
+		height:   *h,
+		colors:   colors,
+		fileName: *fn,
+	}
+
+	// create an empty image
+	img := image.NewRGBA(image.Rectangle{image.Point{0, 0}, image.Point{options.width, options.height}})
+
+	if *isFlag {
+		err, lines = read_lines()
+
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		options.content = processInput(lines, options)
+	}
+
+	// draw func that fill the image with some colors
+	draw(img, options)
+
+	// create a file
+	f, err := os.Create(options.fileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// write an image to a file
+	if err := png.Encode(f, img); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func processInput(lines []string, options imgOptions) [][]*color.RGBA {
+
+	inputArrayWithColorRepresentation := getArray(lines)
+	inputArray := putColors(options, inputArrayWithColorRepresentation)
+	inputArrayUniform := modifyUniform(options, inputArray)
+
+	return scaleContent(options, inputArrayUniform)
+}
+
+func draw(img *image.RGBA, options imgOptions) {
+	// line by line go through the every pixel and fill that with some color
+	var asFlag bool = (options.content != nil)
+	var pixelColor *color.RGBA
+
+	for x := 0; x < (options.width / 2); x++ {
+		for y := 0; y < options.height; y++ {
+			if asFlag {
+				pixelColor = options.content[y][x]
+				img.Set(x, y, pixelColor)
+				pixelColor = options.content[y][options.width-x-1]
+				img.Set(options.width-x-1, y, pixelColor)
+				//
+			} else {
```

Comment:
I would avoid "else" usage. 
https://www.youtube.com/watch?v=5DVV36uqQ4E here is a good explanation about that.

Reply:
+ it's harder to test a func that contains a few 'else's